### PR TITLE
Allow more control over which specs are reused

### LINKED ISSUE #32888: Fine-Grained Control over the reuse-process
### Summary

So, as I mentioned on the Slack, I would like to have all specs with a specific `target=<desc>` at the root be concretized with that target enforced for all dependencies. So I want to disable reusing non-target dependencies I'd argue.

Where should I start? Would I add an option to the concretizer (seems HARD) or do I just override something in lib/spack/spack/spec.py?

Also: it might be sensible to add more specific rules around the reusing-process (consistent compilers, making security-relevant stuff not be reused). Are there any plans for this?

### Rationale

_No response_

### Description

_No response_

### Additional information

_No response_

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

### LINKED ISSUE #42813: Replace `buildable` with `from`
### Summary

Depreciate `buildable: [true, false]` and add `from: [source,  external, buildcache, installed]`

### Rationale

_No response_

### Description

Add `from` along with `buildable`, cannot have both at the same time, `buildable: false` to be replaced with `from: external` 
The order of `from` is up to the discretion of the user

### Additional information

`buildable: false` works the same as `from: [external]`
Should fail if both `buildable` & `from` are used
`from: [source]` without `buildcache` should work similarly to `--fresh`
`from: [source]` without `installed` should work similarly to `--overwrite`

### General information

- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by alalazo**
fixes #42813
fixes #32888

This PR gives users finer control over which specs are reused during concretization.

The value of the `concretizer:reuse` config option now can take an object with the following properties:
- `roots`: true if reusing roots, false if reusing just dependencies
- `exclude`: list of constraints used to select reusable specs 
- `include`: list of constraints used to select reusable specs 
- `from`: allows to select the sources of reused specs

#### Reuse only specs compiled with GCC
```yaml
concretizer:
  reuse:
    roots: true
    include:
    - "%gcc"
```

#### `openmpi` must be used from externals, and it must be the only external used
```yaml
concretizer:
  reuse:
    roots: true
    from:
    - type: local
      exclude:
      - "openmpi"
    - type: buildcache
      exclude:
      - "openmpi"
    - type: external
      include:
      - "openmpi"
```

### Extensions in future PRs

Since the items in the `from:` list are object, they can be extended easily. As an example we might allow something like:
```yaml
concretizer:
  reuse:
    from:
    - type: buildcache
      name: e4s
```
to select specs only from a specific buildcache. Or select specs based on installation dates etc.

### CODE REVIEW

**Comment by alalazo** [comment]
@haampie Changed `reuse:strategy` to `reuse:roots` as we discussed in DMs

**Comment by haampie** [comment]
Can we still change the name of `type: mirror` to `type: buildcache`? In our docs we say

> Build caches refer to mirrors that provide prebuilt packages.

so I think that's more accurate (also corresponds with `--use-buildcache` and `--no-cache` flags from `spack install`.

File: lib/spack/spack/schema/concretizer.py
Comment (psakievich) [inline lib/spack/spack/schema/concretizer.py]: I know I'm late to the review, but how much of a lift would it be to add `upstreams` to this list? Is the intent that those would be coupled under `local`?  I would personally like to see them differentiated if possible.

Comment (alalazo) [reply lib/spack/spack/schema/concretizer.py]: > I would personally like to see them differentiated if possible.

We can have additional PRs later, to add more control - but I'll keep this one as is[^1] :slightly_smiling_face: What would be the use case for distinguishing upstreams from local? 

[^1]: There are also other extensions mentioned in the description that I'd like to add later, for instance selecting a mirror by name, when multiple mirrors are registered.

Comment (psakievich) [reply lib/spack/spack/schema/concretizer.py]: Primarily to have more control. Say I suspect there might be an issue with the upstream, or I wanted to test phasing one out. Follow on is fine by me.

**Comment by alalazo** [comment]
> Can we still change the name of type: mirror to type: buildcache? In our docs we say

See https://github.com/spack/spack/pull/42782/commits/fbc5edb13858aea0e15bcd91f80deef6b1acd308

File: lib/spack/spack/schema/concretizer.py
Comment (tgamblin) [reply lib/spack/spack/schema/concretizer.py]: See #42999 -- a symmetric config option here makes a lot of sense. Upstreams have names and paths that they can be identified by (kind of like mirrors)

File: lib/spack/spack/schema/concretizer.py
Comment (tgamblin) [inline lib/spack/spack/schema/concretizer.py:41]: would really like to be able to write:

```yaml
concretizer:
    reuse:
        from: ["local", "buildcache", "external"]
```

is there an issue w/adding this?

The other thing that is missing from this vs. our idea for `from:` in `packages.yaml` is that there, we were going to specifically allow someone to say `from: "source"`.  How would you say that you *always* want some package to be built from source and never reused here?

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: lib/spack/spack/schema/concretizer.py
Comment (psakievich) [reply lib/spack/spack/schema/concretizer.py:41]: > How would you say that you always want some package to be built from source and never reused here?

Perhaps we could add a global exclusion?
``` yaml
concretizer:
  reuse:
    roots: true
    from:
    - type: local
      exclude:
      - "openmpi"
    - type: buildcache
      exclude:
      - "openmpi"
    - type: external
      include:
      - "openmpi"
    always_exclude:
    - "end_application_a"
    - "end_application_b"
 ```
 or
``` yaml
concretizer:
  reuse:
    roots: true
    from:
    - type: local
      exclude:
      - "openmpi"
    - type: buildcache
      exclude:
      - "openmpi"
    - type: external
      include:
      - "openmpi"
  never_reuse:
    - "end_application_a"
    - "end_application_b"
 ```
 I don't really like my suggested names, but this interface would work for me.

Comment (alalazo) [reply lib/spack/spack/schema/concretizer.py:41]: > How would you say that you always want some package to be built from source and never reused here?

@psakievich @tgamblin There is global exclusion, and imo it's more general than the suggested options:
```yaml
concretizer:
  reuse:
    exclude:
    - "foo"
```
With the above `foo` is not reused from local, buildcache or externals. The only way not to build is to get an accidental hash match. Note that we can also do things like:
```yaml
concretizer:
  reuse:
    exclude:
    - "%oneapi"
```
e.g. rebuild all the specs compiled with `oneapi` etc.

Comment (psakievich) [reply lib/spack/spack/schema/concretizer.py:41]: 🤦‍♂️ well that's what I get for commenting without re-checking the code.

Comment (tgamblin) [reply lib/spack/spack/schema/concretizer.py:41]: ok and `reuse: false` works for everything from source and/or hash?

Comment (tgamblin) [reply lib/spack/spack/schema/concretizer.py:41]: @alalazo: if you do this does it always build from source?  One thing that people had said they wanted was the ability to *always* build a package from source (even if it's installed).  I would not hold this PR up for it but just curious.

`reuse: false` still forces `--fresh` all the time but will do hash reuse, yes?

Comment (alalazo) [reply lib/spack/spack/schema/concretizer.py:41]: > reuse: false still forces --fresh all the time but will do hash reuse, yes?

Yes. Also, in the case above we'll always do hash reuse (i.e. if the final hash is installed already, we won't override it).

An option to always rebuild a spec seems more related to installing, than to reusing specs in concretization :thinking:

Comment (psakievich) [reply lib/spack/spack/schema/concretizer.py:41]: @tgamblin would it need to be **always**?  Like if it is an exact hash match why rebuild?  That would lead to install prefix collisions and would need overwriting etc as I believe @alalazo is alluding to.

 The only exception I can think of is for branch based versions, but that should go away when we add the git-sha pairing to branch versions at concretization. I have a minimum viable product working for that in `spack-manager` that I will need to hook up to the concretizer after we merge #43859.

Comment (tgamblin) [reply lib/spack/spack/schema/concretizer.py:41]: @psakievich People wanted that, but I don't see why we need to hold up the PR over it.

**Review by tgamblin — APPROVED** [review]

**Pull request merged**
